cosi/cosi.go

```diff
@@ -0,0 +1,502 @@
+		}
+	}
+	// add our own commitment to the global V_hat
+	c.aggregateCommitment = c.suite.Point().Add(childVHat, c.commitment)
+	return &Commitment{
+		ChildrenCommit: childVHat,
+		Commitment:     c.commitment,
+	}
+
+}
+
+// CreateChallenge creates the challenge out of the message it has been given.
+// This is typically called by Root.
+func (c *Cosi) CreateChallenge(msg []byte) (*Challenge, error) {
+	// H( Commit || AggPublic || M)
+	hash := sha512.New()
+
+	pb, err := c.aggregateCommitment.MarshalBinary()
+	if err != nil {
+		return nil, err
+	}
+	hash.Write(pb)
+	pb, err = c.mask.Aggregate().MarshalBinary()
+	if err != nil {
+		return nil, err
+	}
+	hash.Write(pb)
+	hash.Write(msg)
+	chalBuff := hash.Sum(nil)
+	// reducing the challenge
+	c.challenge = sliceToSecret(c.suite, chalBuff)
+	c.message = msg
+	return &Challenge{
+		Challenge:        c.challenge,
+		GlobalCommitment: c.aggregateCommitment,
+	}, nil
+}
+
+// Challenge keeps in memory the Challenge from the message.
+func (c *Cosi) Challenge(ch *Challenge) *Challenge {
+	c.challenge = ch.Challenge
+	c.globalCommitment = ch.GlobalCommitment
+	return ch
+}
+
+// CreateResponse is called by a leaf to create its own response from the
+// challenge + commitment + private key. It returns the response to send up to
+// the tree.
+func (c *Cosi) CreateResponse() (*Response, error) {
+	err := c.genResponse()
+	return &Response{Response: c.response}, err
+}
+
+// Response generates the response from the commitment, challenge and the
+// responses of its children.
+func (c *Cosi) Response(responses []*Response) (*Response, error) {
+	// //create your own response
+	if err := c.genResponse(); err != nil {
+		return nil, err
+	}
+	aggregateResponse := c.suite.Secret().Zero()
+	for _, resp := range responses {
+		// add responses of child
+		aggregateResponse = aggregateResponse.Add(aggregateResponse, resp.Response)
+		// add responses of it's children if there is one (i.e. if it is not a
+		// leaf)
+		if resp.ChildrenResp != nil {
+			aggregateResponse = aggregateResponse.Add(aggregateResponse, resp.ChildrenResp)
+		}
+	}
+	// Add our own
+	c.aggregateResponse = c.suite.Secret().Add(aggregateResponse, c.response)
+	return &Response{
+		Response:     c.response,
+		ChildrenResp: aggregateResponse,
+	}, nil
+}
+
+// Signature returns a signature in the like the EdDSA signature format plus the
+// bitmask at the end.
+// Sig = AggCommit || AggResponse || bitmask
+func (c *Cosi) Signature() []byte {
+	// Sig = R || S || bitmask
+	sigS := secretToSlice(c.aggregateResponse)
+	sigR, err := c.aggregateCommitment.MarshalBinary()
+	if err != nil {
+		panic("Can't generate signature !")
+	}
+	final := make([]byte, 64+c.mask.MaskLen())
+	copy(final[:], sigR)
+	copy(final[32:64], sigS)
+	copy(final[64:], c.mask.mask)
+	return final
+}
+
+// GetAggregateResponse returns the aggregated response that this cosi has
+// accumulated.
+func (c *Cosi) GetAggregateResponse() abstract.Secret {
```

Comment:
Yes, I know the convention but for the case of `GetChallenge()` and `GetCommitment()` I can't remove the -Get- part because a function already exists without the -Get-.  I wanted to stay consistent with the other getters here also and not having a -Get-.  Maybe the name of `Challenge` and `Commitment` is not adequate then, do any of you have other ideas ?
`ChallengePhase()` / `CommitmentPhase` ? Not elegant but explicit ...